client/network/src/discovery.rs

```diff
@@ -547,14 +589,13 @@ impl NetworkBehaviour for DiscoveryBehaviour {
 
 	fn inject_connection_closed(
 		&mut self,
-		peer_id: &PeerId,
-		conn: &ConnectionId,
-		endpoint: &ConnectedPoint,
+		_peer_id: &PeerId,
+		_conn: &ConnectionId,
+		_endpoint: &ConnectedPoint,
+		_handler: <Self::ProtocolsHandler as IntoProtocolsHandler>::Handler,
 	) {
 		self.num_connections -= 1;
-		for k in self.kademlias.values_mut() {
-			NetworkBehaviour::inject_connection_closed(k, peer_id, conn, endpoint)
-		}
+		// NetworkBehaviour::inject_connection_closed on Kademlia<MemoryStore> does nothing.
```

Comment:
That's not really future-proof. Maybe in a future version it will do something.
We need a way to deconstruct a `MultiHandler` and libp2p doesn't provide any. When this PR is merged, could you open an issue in libp2p and in Substrate and follow up with a fix?

Reply:
@mxinden can you take a look at this one? ^

Reply:
> We need a way to deconstruct a MultiHandler and libp2p doesn't provide any. 

Any road blockers to contributing this upstream (rust-libp2p)?